### ISSUE TITLE
Reproducing Hallway

### ISSUE THREAD

**Issue by liquidnode** [vub-ai-lab/bdpi#2]
I'm currently trying to reproduce the Hallway experiment from the paper and I found some inconsistencies. First I noticed that the average reward converges to 0.7 instead of the 0.95 from the paper (below are the last 10 lines of my output file), so the policy at the end of the 500 epsiodes is far from optimal. The only difference I made to the paper is that I set loops=8 and actor-count=8 instead of 16, altough in the paper this seems to have no impact on the training of the agent. I also set the learning rate to 1e-5 as it's done in experiments_gym.sh, which was not specified in the paper. I will start a new identical experiment to confirm that the average reward converges against 0.7, but note that I'm limited in compute resources so this might take a while.

Last 10 lines of the output file:
0.0 0.682810541837494 63985 65488.765096 250
0.956 0.685542436419119 64040 65545.814269 55
0.9424 0.6881110120549279 64112 65621.469044 72
0.9432 0.6906619019343786 64183 65695.824354 71
0.9872 0.6936272829150348 64199 65712.412977 16
0.9264 0.6959550100858846 64291 65808.075586 92
0.9128000000000001 0.6981234599850258 64400 65921.449253 109
0.9536 0.7006782253851755 64458 65982.167581 58
0.0 0.6936714431313238 64708 66242.778081 250
0.0 0.6867347287000105 64958 66502.851716 250

I also found that the enviroment described in the paper might not be the same as the one found in [gym-miniworld](https://github.com/maximecb/gym-miniworld). There they add a small penalty to the rewards depending on the number of steps the agent needs to reach the goal. This penalty is the difference between an almost perfect and a good policy trained by BDPI in regards to the training graphs in the paper. MiniWorld-Hallway-v0 is also not as random as stated in the paper. Only the starting position of the agent is randomized. Is the enviroment in the paper a different one or was it an older version of MiniWorld-Hallway-v0?

**Comment by steckdenis** [comment]
Hello,

I've looked at the commit history of Hallway, and I don't see anything that may have changed in any meaningful way between master and the version I used. I'm looking for the exact commit I use in my experiments. This was a checkout of gym-miniworld (https://github.com/maximecb/gym-miniworld) from about January.

Regarding the results, the results reproduced by the paper are exactly as produced by "./experiments_gym.sh hallway && cat commands_hallway.sh | parallel". They indeed use a learning rate of 1e-5, 16 critics, 16 loops, 4 qloops and 1 epochs. To be more precise, the Hallway curve is generated with the out-parallel-hallway-rp-16critics-er256-epochs1x16-qloops4-* files produces by the experiments.

For the difference that you observe, I see two possible reasons:

- The first one is the amount of runs. The line you show all come from a single run, where the agent sometimes achieves around 0.93, and sometimes misses the goal. This is an "unlucky" run, in which the agent did not learn an optimal policy. If you perform 8 runs, as done in the paper, you should have many "lucky" runs where the agent manages to obtain 0.93-0.95 all the time.
- Decreasing the amount of critics should not have an impact, as shown in the paper. But we never know. Maybe the proportion of unlucky runs slightly increases when the number of critics is decreased?

Regarding the randomization of the room, I have a distant memory of having read a line that causes the length of the room to be random. I couldn't find the line in https://github.com/maximecb/gym-miniworld/blob/master/gym_miniworld/envs/hallway.py anymore. I'll look at that in more detail.

**Comment by liquidnode** [comment]
Thanks for your reply. I'll start a new run with exactly the same parameters stated in the paper. It could be an unlucky run I experienced, though it is still performing way better than PPO, so this is "a complaint on a high comfort level". Although I would expect that the training still converges to the same average reward in the end, because it should be quite unlikely that the training of all Q-networks gets stuck in a local minima.

About the randomization: I digged a bit more in the source files of gym-miniwold and it seems that the agent and box placement is randomized. Only the size of the room isn't.

**Comment by steckdenis** [comment]
I'm interested in the results you obtain :) . BDPI has been extremely consistent on my machine, producing almost the same results run after run. I would be curious to see if this can be reproduced, or if there is something with my setting that matters (I compile PyTorch myself and run it on my CPU, a 32-core Threadripper 2990WX that is faster than a GPU for many RL tasks).

**Comment by liquidnode** [comment]
Ok results for one run are in.
![plot](https://user-images.githubusercontent.com/19669979/61992605-e7928900-b060-11e9-82c1-ec30f3baca2c.png)
The blue curve is produced by the rewards of the enviroment directly. Since the positive rewards are always slightly lower than 1, I set positive rewards to 1 in the orange curve to have a better comparison to the paper. 

This time the results are more plausible than the first time, though it still underperforms by a small margin compared to the paper. I think the reason is twofold. First the reward signal is more stochastic than in the paper, since it can not know how much steps it took before arriving at the goal. And second I noticed that the maximal episode length was halfed to 250 steps, so the agent does not have as much time to find the goal as compared to the paper and training time at the start is essentially halfed. So I guess to perfectly reproduce the results in the paper, one has to modify the rewards and increase the episode length, which is unfortunately not accessible in Hallway. To solve this one could revert to before https://github.com/maximecb/gym-miniworld/commit/39cc95b5a70950cdfb4ce5fd70c9e1be0a38ce57#diff-7738633d95bd718a4fee5eb306812581, though these results are close enough for me, so I would close this issue if there are no objections from your side.

**Comment by steckdenis** [comment]
Wouah, impressive. I'm very happy that you found the commit that changes the episode length (it is well hidden in the commit). I will update the README file and the paper in this repository with this information. I think that the main reason for the difference is the episode length. In the paper, I don't mention the time-step penalty, but this may be because I missed it (there is no full description of the Miniworld environments, unfortunately, so I had to read the code and describe what I saw in the paper).

Thank you again for your investigation ! :)
